File: java/src/main/java/problems101_200/MaximumDepthOfBinaryTree.java
package problems101_200;

import concepts.TreeNode;

enum MaximumDepthOfBinaryTree {
    DEFAULT {
        @Override
        int maxDepth(TreeNode root) {
            return recursiveMaxDepth(root, 0);
        }

        int recursiveMaxDepth(TreeNode node, int depth) {
            if (node == null) {
                return depth;
            }
            int leftDepth = recursiveMaxDepth(node.left, depth + 1);
            int rightDepth = recursiveMaxDepth(node.right, depth + 1);
            return Math.max(leftDepth, rightDepth);
        }
    };

    abstract int maxDepth(TreeNode root);
}

File: kotlin/src/main/kotlin/problems101_200/MaximumDepthOfBinaryTree.kt
package problems101_200

import concepts.TreeNode
import kotlin.math.max

enum class MaximumDepthOfBinaryTree {
    DEFAULT {
        override fun maxDepth(root: TreeNode?): Int = recursiveMaxDepth(root, 0)

        fun recursiveMaxDepth(node: TreeNode?, depth: Int): Int {
            if (node == null) {
                return depth
            }
            val leftDepth = recursiveMaxDepth(node.left, depth + 1)
            val rightDepth = recursiveMaxDepth(node.right, depth + 1)
            return max(leftDepth, rightDepth)
        }
    },
    ;

    abstract fun maxDepth(root: TreeNode?): Int
}
